### ISSUE TITLE
Resource with typed member variable or a typed return produces Orphan instance at exit

### ISSUE THREAD

**Issue by MikeSchulze** [godotengine/godot#46467]
**Godot version:**
Godot Engine v3.2.4.beta3.official

**OS/device including version:**
windows 10

**Issue description:**
I run into this issue when running a script from command line. It shows this error message at exit.
```
ERROR: ~List: Condition "_first != __null" is true.
   At: ./core/self_list.h:112
ERROR: ~List: Condition "_first != __null" is true.
   At: ./core/self_list.h:112
WARNING: cleanup: ObjectDB instances leaked at exit (run with --verbose for details).
     At: core/object.cpp:2132
Leaked instance: GDScriptNativeClass:536
Leaked instance: GDScript:1160 - Resource path: res://TestResource.gd
Hint: Leaked instances typically happen when nodes are removed from the scene tree (with `remove_child()`) but not freed (with `free()` or `queue_free()`).
ERROR: clear: Resources still in use at exit (run with --verbose for details).
   At: core/resource.cpp:475
Resource still in use: res://TestResource.gd (GDScript)
Orphan StringName: res://TestResource.gd
Orphan StringName: parent
Orphan StringName: GDScriptNativeClass
Orphan StringName: _name
Orphan StringName: res://TestResource.gd::13::TestResource.set_parent
Orphan StringName: res://TestResource.gd::10::TestResource._init
Orphan StringName: res://TestResource.gd::16::TestResource.name
Orphan StringName: set_parent
Orphan StringName: res://TestResource.gd::19::TestResource.parent
Orphan StringName: _parent
Orphan StringName: name
Orphan StringName: Resource
Orphan StringName: _init
Orphan StringName: GDScript
StringName: 14 unclaimed string names at exit.
```

This error occures when i use a typed member on a resource class where should be freeing by default if not in use.
Using a untyped member no error is shown. 
````
class_name TestResource
extends Resource

# if using typed var than this line produces the error
# ERROR: clear: Resources still in use at exit
var _parent :TestResource = null

...
````

**Steps to reproduce:**
* use the attached example project 
* open windows command console and run 'runScript.bat'

**Minimal reproduction project:**
[orphanResources.zip](https://github.com/godotengine/godot/files/6053964/orphanResources.zip)

**Comment by MikeSchulze** [comment]
It produces also orphans when use a typed return function
```
class_name TestClass
extends Resource


const WARNINGS = "warnings"
const FAILED = "failed"
const ERRORS = "errors"


enum  {
	INIT,
	STOP,
}

func before(suite_name :String, total_count) -> TestClass:
	return self

```

```
Loading resource: res://./addons/gdUnit3/GdTest.gd
Loading resource: res://addons/gdUnit3/src/core/event/TestClass.gd
Loading resource: res://default_env.tres
Loaded builtin certs
ERROR: ~List: Condition "_first != __null" is true.
   At: ./core/self_list.h:112.
ERROR: ~List: Condition "_first != __null" is true.
   At: ./core/self_list.h:112.
WARNING: cleanup: ObjectDB instances leaked at exit (run with --verbose for details).
   At: core/object.cpp:2135.
Leaked instance: GDScript:1160 - Resource path: res://addons/gdUnit3/src/core/event/TestClass.gd
Leaked instance: GDScriptNativeClass:538
Hint: Leaked instances typically happen when nodes are removed from the scene tree (with `remove_child()`) but not freed (with `free()` or `queue_free()`).
ERROR: clear: Resources still in use at exit (run with --verbose for details).
   At: core/resource.cpp:477.
Resource still in use: res://addons/gdUnit3/src/core/event/TestClass.gd (GDScript)
Orphan StringName: total_count
Orphan StringName: STOP
Orphan StringName: before
Orphan StringName: ERRORS
Orphan StringName: res://addons/gdUnit3/src/core/event/TestClass.gd::15::TestClass.before
Orphan StringName: res://addons/gdUnit3/src/core/event/TestClass.gd
Orphan StringName: GDScriptNativeClass
Orphan StringName: WARNINGS
Orphan StringName: res://addons/gdUnit3/src/core/event/TestClass.gd::0::TestClass._init
Orphan StringName: suite_name
Orphan StringName: INIT
Orphan StringName: Resource
Orphan StringName: FAILED
Orphan StringName: _init
Orphan StringName: GDScript
StringName: 15 unclaimed string names at exit.
```

That is really annoying.
I get now tons of orphan error since i try to execute my project as an CLI tool.
I want to use typed scripts but this bug makes this imposible.

This is an critical issue in my eyes and needs to be fixed as soon as possible.
And please don't say fix only in 4.0.

**Comment by HaSa1002** [comment]
Not reproducible in 4.0, likely because of the new GDScript parser.

<details>
<summary>Details</summary>

```
PS D:\Software\godot\bin> .\godot.windows.tools.64.exe -d -s "F:\Godot\bugs\orphanResources_4\OrphanResources.gd" --verbose
Godot Engine v4.0.dev.custom_build.4de0768cd - https://godotengine.org
Using "ICU / HarfBuzz / Graphite" text server...
WARNING: GENERAL - Message Id Number: 0 | Message Id Name: Loader Message
        ReadDataFilesInRegistry: Registry lookup failed to get layer manifest files.
        Objects - 1
                Object[0] - VK_OBJECT_TYPE_INSTANCE, Handle 2688811372368
     at: VulkanContext::_debug_messenger_callback (drivers\vulkan\vulkan_context.cpp:152)
Using Vulkan Device #0: AMD - Radeon RX 580 Series
Using "wintab" pen tablet driver...
WASAPI: wFormatTag = 65534
WASAPI: nChannels = 2
WASAPI: nSamplesPerSec = 48000
WASAPI: nAvgBytesPerSec = 384000
WASAPI: nBlockAlign = 8
WASAPI: wBitsPerSample = 32
WASAPI: cbSize = 22
WASAPI: detected 2 channels
WASAPI: audio buffer frames: 1962 calculated latency: 44ms

CORE API HASH: 5933855955601296711
EDITOR API HASH: 11625017473697416794

Debugger Break, Reason: 'Parser Error: Identifier "TestResource" not declared in the current scope.'
*Frame 0 - F:/Godot/bugs/orphanResources_4/OrphanResources.gd:5 in function ''
Enter "help" for assistance.
debug> c
SCRIPT ERROR: Parse Error: Identifier "TestResource" not declared in the current scope.
          at: GDScript::reload (F:/Godot/bugs/orphanResources_4/OrphanResources.gd:5)
ERROR: Method/function failed. Returning: ERR_PARSE_ERROR
   at: GDScript::reload (modules\gdscript\gdscript.cpp:832)
PS D:\Software\godot\bin> .\godot.windows.tools.64.exe -d -s "F:\Godot\bugs\orphanResources_4\OrphanResources.gd" --verbose
Godot Engine v4.0.dev.custom_build.4de0768cd - https://godotengine.org
Using "ICU / HarfBuzz / Graphite" text server...
WARNING: GENERAL - Message Id Number: 0 | Message Id Name: Loader Message
        ReadDataFilesInRegistry: Registry lookup failed to get layer manifest files.
        Objects - 1
                Object[0] - VK_OBJECT_TYPE_INSTANCE, Handle 1682490912464
     at: VulkanContext::_debug_messenger_callback (drivers\vulkan\vulkan_context.cpp:152)
Using Vulkan Device #0: AMD - Radeon RX 580 Series
Using "wintab" pen tablet driver...
WASAPI: wFormatTag = 65534
WASAPI: nChannels = 2
WASAPI: nSamplesPerSec = 48000
WASAPI: nAvgBytesPerSec = 384000
WASAPI: nBlockAlign = 8
WASAPI: wBitsPerSample = 32
WASAPI: cbSize = 22
WASAPI: detected 2 channels
WASAPI: audio buffer frames: 1962 calculated latency: 44ms

CORE API HASH: 5933855955601296711
EDITOR API HASH: 11625017473697416794

Debugger Break, Reason: 'Parser Error: Identifier "TestResource" not declared in the current scope.'
*Frame 0 - F:/Godot/bugs/orphanResources_4/OrphanResources.gd:5 in function ''
Enter "help" for assistance.
debug> q
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] SceneTree::quit (D:\Software\godot\scene\main\scene_tree.cpp:539)
[1] LocalDebugger::debug (D:\Software\godot\core\debugger\local_debugger.cpp:277)
[2] ScriptDebugger::debug (D:\Software\godot\core\debugger\script_debugger.cpp:100)
[3] GDScriptLanguage::debug_break_parse (D:\Software\godot\modules\gdscript\gdscript_editor.cpp:240)
[4] GDScript::reload (D:\Software\godot\modules\gdscript\gdscript.cpp:824)
[5] GDScriptCache::get_full_script (D:\Software\godot\modules\gdscript\gdscript_cache.cpp:211)
[6] ResourceFormatLoaderGDScript::load (D:\Software\godot\modules\gdscript\gdscript.cpp:2325)
[7] ResourceLoader::_load (D:\Software\godot\core\io\resource_loader.cpp:194)
[8] ResourceLoader::_thread_load_function (D:\Software\godot\core\io\resource_loader.cpp:221)
[9] ResourceLoader::load (D:\Software\godot\core\io\resource_loader.cpp:578)
[10] Main::start (D:\Software\godot\main\main.cpp:1992)
[11] widechar_main (D:\Software\godot\platform\windows\godot_windows.cpp:161)
[12] _main (D:\Software\godot\platform\windows\godot_windows.cpp:185)
[13] main (D:\Software\godot\platform\windows\godot_windows.cpp:199)
[14] __scrt_common_main_seh (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[15] BaseThreadInitThunk
-- END OF BACKTRACE --
```

</details>

**Comment by HaSa1002** [comment]
Not reproducible in 3.2 main, too.

<details>
<summary>Details</summary>

```
PS D:\Software\godot32\bin> .\godot.windows.tools.64.exe -d -s "F:\Godot\bugs\orphanResources\OrphanResources.gd" --verbose
Godot Engine v3.2.4.rc.custom_build.8a020a657 - https://godotengine.org
Using GLES3 video driver
OpenGL ES 3.0 Renderer: Radeon RX 580 Series
OpenGL ES Batching: ON
        OPTIONS
        max_join_item_commands 16
        colored_vertex_format_threshold 0.25
        batch_buffer_size 16384
        light_scissor_area_threshold 1
        item_reordering_lookahead 4
        light_max_join_items 32
        single_rect_fallback False
        debug_flash False
        diagnose_frame False
WASAPI: wFormatTag = 65534
WASAPI: nChannels = 2
WASAPI: nSamplesPerSec = 48000
WASAPI: nAvgBytesPerSec = 384000
WASAPI: nBlockAlign = 8
WASAPI: wBitsPerSample = 32
WASAPI: cbSize = 22
WASAPI: detected 2 channels
WASAPI: audio buffer frames: 1962 calculated latency: 44ms

CORE API HASH: 4754945593587898049
EDITOR API HASH: 13330850812510104001
Loading resource: F:/Godot/bugs/orphanResources/OrphanResources.gd

Debugger Break, Reason: 'Parser Error: The identifier "TestResource" isn't declared in the current scope.'
*Frame 0 - F:/Godot/bugs/orphanResources/OrphanResources.gd:5 in function ''
Enter "help" for assistance.
debug> c
SCRIPT ERROR: GDScript::reload: Parse Error: The identifier "TestResource" isn't declared in the current scope.
          At: F:/Godot/bugs/orphanResources/OrphanResources.gd:5
ERROR: GDScript::reload: Method failed. Returning: ERR_PARSE_ERROR
   At: modules\gdscript\gdscript.cpp:583
PS D:\Software\godot32\bin> .\godot.windows.tools.64.exe -d -s "F:\Godot\bugs\orphanResources\OrphanResources.gd" --verbose
Godot Engine v3.2.4.rc.custom_build.8a020a657 - https://godotengine.org
Using GLES3 video driver
OpenGL ES 3.0 Renderer: Radeon RX 580 Series
OpenGL ES Batching: ON
        OPTIONS
        max_join_item_commands 16
        colored_vertex_format_threshold 0.25
        batch_buffer_size 16384
        light_scissor_area_threshold 1
        item_reordering_lookahead 4
        light_max_join_items 32
        single_rect_fallback False
        debug_flash False
        diagnose_frame False
WASAPI: wFormatTag = 65534
WASAPI: nChannels = 2
WASAPI: nSamplesPerSec = 48000
WASAPI: nAvgBytesPerSec = 384000
WASAPI: nBlockAlign = 8
WASAPI: wBitsPerSample = 32
WASAPI: cbSize = 22
WASAPI: detected 2 channels
WASAPI: audio buffer frames: 1962 calculated latency: 44ms

CORE API HASH: 4754945593587898049
EDITOR API HASH: 13330850812510104001
Loading resource: F:/Godot/bugs/orphanResources/OrphanResources.gd

Debugger Break, Reason: 'Parser Error: The identifier "TestResource" isn't declared in the current scope.'
*Frame 0 - F:/Godot/bugs/orphanResources/OrphanResources.gd:5 in function ''
Enter "help" for assistance.
debug> q
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] SceneTree::quit (D:\Software\godot32\scene\main\scene_tree.cpp:647)
[1] ScriptDebuggerLocal::debug (D:\Software\godot32\core\script_debugger_local.cpp:214)
[2] GDScriptLanguage::debug_break_parse (D:\Software\godot32\modules\gdscript\gdscript_editor.cpp:230)
[3] GDScript::reload (D:\Software\godot32\modules\gdscript\gdscript.cpp:580)
[4] ResourceFormatLoaderGDScript::load (D:\Software\godot32\modules\gdscript\gdscript.cpp:2246)
[5] ResourceLoader::_load (D:\Software\godot32\core\io\resource_loader.cpp:270)
[6] ResourceLoader::load (D:\Software\godot32\core\io\resource_loader.cpp:384)
[7] Main::start (D:\Software\godot32\main\main.cpp:1651)
[8] widechar_main (D:\Software\godot32\platform\windows\godot_windows.cpp:160)
[9] _main (D:\Software\godot32\platform\windows\godot_windows.cpp:184)
[10] main (D:\Software\godot32\platform\windows\godot_windows.cpp:196)
[11] __scrt_common_main_seh (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[12] BaseThreadInitThunk
-- END OF BACKTRACE --
```

</details>

**Comment by MikeSchulze** [comment]
your reports shows some crashes!

**Comment by HaSa1002** [comment]
Yeah, but not related to this issue. The debugger crashes on quit and I believe this is intended

**Comment by HaSa1002** [comment]
I executed your script twice. The first time (when I continue execution), you only get a parser error.

**Comment by MikeSchulze** [comment]
hm ok you used latest v3.2.x build right?
I will try with nightly build, give me some minutes ;)

**Comment by MikeSchulze** [comment]
Very NICE!!!!!
I take some small tests on my Mac based on (Godot_v3.2.4-rc2_osx.universal.zip) and it works. 

That saves my day, i'm happy.
Thanks

**Comment by HaSa1002** [comment]
We get the crash, because godot only executes the script. This might also lead to the parser error (I guess). I am not familiar with the command line options. Can't you run the whole project and start "OrphanResources.gd" attached to a tree.

**Comment by MikeSchulze** [comment]
So i guess this can be closed as solved with V3.2.4_rc2 :)

**Comment by HaSa1002** [comment]
I think so.
